jupyter_server/auth/identity.py

```diff
@@ -0,0 +1,155 @@
+"""Identity Provider interface
+
+This defines the _authentication_ layer of Jupyter Server,
+to be used in combination with Authorizer for _authorization_.
+
+.. versionadded:: 2.0
+"""
+import sys
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+
+from tornado.web import RequestHandler
+from traitlets.config import LoggingConfigurable
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict
+else:
+    try:
+        from typing_extensions import TypedDict
+    except ImportError:
+        TypedDict = Dict
+
+
+class IdentityModel(TypedDict):
+    # see the JupyterLab IUser model for definitions
+
+    username: str  # the only truly required field
+
+    # these fields are derived from username if not specified
+    name: str
+    display_name: str
+
+    # these fields are left as None if undefined
+    initials: Optional[str]
+    avatar_url: Optional[str]
+    color: Optional[str]
+
+    # Jupyter Server permissions
+    # as a dict of permitted {"resource": ["actions"]}
+    permissions: Dict[str, List[str]]
+
+
+class IdentityProvider(LoggingConfigurable):
+    """
+    Interface for providing identity
+
+    Two principle methods:
+
+    - :meth:`~.IdentityProvider.get_user` returns a user object.
+      For successful authentication,
+      this may return anything truthy.
+
+    - :meth:`~.IdentityProvider.identity_model` returns a standard identity model dictionary,
+      for use in the /me API.
+      This should accept whatever is returned from get_user()
+      and return a dictionary matching the structure of
+      :class:`~.IdentityModel`.
+
+    .. versionadded:: 2.0
+    """
+
+    def get_user(self, handler: RequestHandler) -> Any:
+        """Get the authenticated user for a request
+
+        User may be anything truthy, but must be understood by identity_model method.
+
+        Return None if the request is not authenticated.
+
+        When in doubt, use a standard identity model.
+        """
+
+        if handler.login_handler is None:
+            return {
+                "username": "anonymous",
+            }
+
+        # The default: call LoginHandler.get_user for backward-compatibility
+        # TODO: move default implementation to this class,
+        # deprecate `LoginHandler.get_user`
+        user = handler.login_handler.get_user(handler)
+        return user
+
+    def identity_model(self, user: Any) -> IdentityModel:
+        """Construct standardized identity model for the identity API
+
+        Casts objects returned by `.get_user` (generally str username or dict with 'username' or 'name')
+        To a complete IdentityModel dict.
+
+        `username` is required.
+        Any other missing fields will be filled out with defaults.
+        """
+        identity = {}
+        if isinstance(user, str):
+            return {
+                "username": user,
+            }
```

Comment:
Is this meant to be filled as well? There seems to be some ambiguity between this behavior, the docstring, and the need for the private method `_get_identity_model`.